pkg/sql/schemachanger/scgraph/graph.go

```diff
@@ -249,3 +262,101 @@ func (de *DepEdge) To() *scpb.Node { return de.to }
 
 // Name returns the name of the rule which generated this edge.
 func (de *DepEdge) Name() string { return de.rule }
+
+// GetNodeRanks fetches ranks of nodes in topological order.
+func (g *Graph) GetNodeRanks() map[*scpb.Node]int {
+	backCycleExists := func(n *scpb.Node, de *DepEdge) bool {
+		var foundBack bool
+		_ = g.ForEachDepEdgeFrom(de.To(), func(maybeBack *DepEdge) error {
+			foundBack = foundBack || maybeBack.To() == n
+			return nil
+		})
+		return foundBack
+	}
+	l := list.New()
+	marks := make(map[*scpb.Node]bool)
+	var visit func(n *scpb.Node)
+	visit = func(n *scpb.Node) {
+		permanent, marked := marks[n]
+		if marked && !permanent {
+			panic("not a dag")
```

Comment:
Thoughts on catching this panic and converting it to an error and then having this method return an error?